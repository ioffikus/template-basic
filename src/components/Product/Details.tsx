import React from 'react';
import { connect } from 'react-redux';
import { Tabs, Table } from 'antd';
import { Link } from 'react-router-dom';
import _ from 'lodash';
import Price from 'src/components/Price';

interface IOwnProps {}

interface IConnectedState {
  ui: Alicanto.Models.UI;
  uid: string;
  schemaFields: Alicanto.Models.CatalogItemSchemaFields;
  categories: Alicanto.Models.Category[];
  modifications: Alicanto.Models.CatalogItem[];
  isBundle: boolean;
  bundleContent: Alicanto.Models.CatalogItem[];
}

type IProps = IOwnProps & IConnectedState;

interface IState {
  activeKey: string;
  specificationsData: { key: string; label: string; value: string }[];
}

const mapStateToProps = (state: Store.IState, ownProps: IOwnProps) => ({
  ui: state.ui,
  uid: state.product.uid,
  schemaFields: state.product.schema_fields,
  categories: state.product.categories, // TODO: design is not released yet
  modifications: state.product.modifications, // TODO: design is not released yet
  isBundle: state.product.is_bundle, // TODO: design is not released yet
  bundleContent: state.product.bundle_content, // TODO: design is not released yet
});

const WrappedDetails = connect<IConnectedState, undefined, IOwnProps>(mapStateToProps)(
  class Details extends React.Component<IProps, IState> {
    readonly defaultKey = 'specifications';

    readonly specificationsColumns = [{ dataIndex: 'label' }, { dataIndex: 'value' }];

    state: IState = {
      specificationsData: this.getSpecificationsData(this.props.schemaFields),
      activeKey: this.defaultKey,
    };

    componentWillReceiveProps(nextProps: IProps) {
      if (this.props.uid !== nextProps.uid) {
        const specificationsData = this.getSpecificationsData(nextProps.schemaFields);
        this.setState({ specificationsData, activeKey: this.defaultKey });
      }
    }

    handleChange = (activeKey: string) => {
      this.setState({ activeKey });
    };

    getSpecificationsData(schemaFields: Alicanto.Models.CatalogItemSchemaFields) {
      const result: { key: string; label: string; value: string }[] = [];
      Object.keys(schemaFields).forEach(key => {
        if (key === 'picture_url') {
          return;
        }
        const schemaField = _.get(schemaFields, key);
        const inList = _.get(schemaField, 'in_list', null);
        if (!_.isNull(inList)) {
          const label = _.get(schemaField, 'label', '');
          const value = _.get(schemaField, 'value', '');
          result.push({ key, label, value });
        }
      });
      return result;
    }

    render() {
      return (
        <div className="root">
          <style jsx>{`
            .root :global(.ant-tabs-bar) {
              text-transform: uppercase;
            }
            .root :global(.ant-tabs-tabpane) {
              padding-left: 26px;
              padding-right: 26px;
            }
            .root :global(.ant-table) {
              font-size: 14px;
            }
            .root :global(.ant-table-tbody > tr > td) {
              padding-top: 17px;
              padding-bottom: 17px;
            }
            .root :global(.ant-table td:first-of-type) {
              color: #b2bdcb;
            }
            .root :global(.ant-table tr:hover > td) {
              background: transparent;
            }
            li {
              padding-top: 17px;
              padding-bottom: 17px;
            }
            .root :global(.bundle),
            .root :global(.modification),
            .root :global(.category) {
              font-size: 14px;
            }
            .root :global(.modification .currency) {
              font-size: inherit;
              color: inherit;
            }
          `}</style>
          <Tabs activeKey={this.state.activeKey} onChange={this.handleChange}>
            <Tabs.TabPane tab="Specifications" key="specifications">
              <Table
                columns={this.specificationsColumns}
                dataSource={this.state.specificationsData}
                pagination={false}
                showHeader={false}
              />
            </Tabs.TabPane>
            {!_.isEmpty(this.props.bundleContent) && (
              <Tabs.TabPane tab="Bundle" key="bundle">
                <ul>
                  {this.props.bundleContent.map(bundle => {
                    const schemaFields = bundle.schema_fields;
                    const shortDescription = _.get(schemaFields, 'short_description.value', '');
                    const name = _.get(schemaFields, 'name.value', '');
                    return (
                      <li key={bundle.uid}>
                        <Link to={`/product/${bundle.uid}`} title={shortDescription} className="bundle">
                          {name}
                        </Link>
                      </li>
                    );
                  })}
                </ul>
              </Tabs.TabPane>
            )}
            {!_.isEmpty(this.props.modifications) && (
              <Tabs.TabPane tab="Modifications" key="modifications">
                <ul>
                  {this.props.modifications.map(modification => {
                    const schemaFields = modification.schema_fields;
                    const shortDescription = _.get(schemaFields, 'short_description.value', '');
                    const name = _.get(schemaFields, 'name.value', '');
                    const currency: string = _.get(schemaFields, 'currency.value', '');
                    const price = _.get(schemaFields, 'price.value', 0);
                    return (
                      <li key={modification.uid}>
                        <Link to={`/product/${modification.uid}`} title={shortDescription} className="modification">
                          {name} -&nbsp;
                          <Price price={price} currency={currency} />
                        </Link>
                      </li>
                    );
                  })}
                </ul>
              </Tabs.TabPane>
            )}
            {!_.isEmpty(this.props.categories) && (
              <Tabs.TabPane tab="Categories" key="categories">
                <ul>
                  {this.props.categories.map(category => {
                    return (
                      <li key={category.uid}>
                        <Link to={`/catalog/${category.uid}`} className="category">
                          {category.name}
                        </Link>
                      </li>
                    );
                  })}
                </ul>
              </Tabs.TabPane>
            )}
          </Tabs>
        </div>
      );
    }
  },
);

export default WrappedDetails;
