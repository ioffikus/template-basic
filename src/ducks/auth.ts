import { createAction, createReducer } from 'redux-act';
import { AxiosResponse } from 'axios';
import { push } from 'connected-react-router';
import { GET, POST } from 'src/core/API';
import EndpointsConfig from 'src/configs/endpoints';

export const REDUCER = 'AUTH';
const NS = `${REDUCER}__`;

export interface IAuthState extends Store.IReduxCommonState {
  isAuthenticated: boolean;
  isAuthLinkRequesting: boolean;
  url: string;
  data: object;
}

export const initialState: IAuthState = {
  isAuthenticated: false,
  isRequesting: false,
  isRequestError: false,
  isAuthLinkRequesting: false,
  url: null,
  data: null,
};

const reducer = createReducer<IAuthState>({}, initialState);

export const loginUserSuccess = createAction(`${NS}LOGIN_USER_SUCCESS`);
reducer.on(loginUserSuccess, state => ({
  ...state,
  isRequesting: false,
  isRequestError: false,
  isAuthenticated: true,
}));

const authLinkSuccess = createAction<{ url: string; data: object }>(`${NS}AUTH_LINK_SUCCESS`);
reducer.on(authLinkSuccess, (state, { url, data }) => ({
  ...state,
  url,
  data: { ...data },
  isAuthLinkRequesting: false,
}));

const loginUserFailure = createAction(`${NS}LOGIN_USER_FAILURE`);
reducer.on(loginUserFailure, state => ({
  ...state,
  isRequesting: false,
  isAuthenticated: false,
  isRequestError: true,
}));

const authLinkFailure = createAction(`${NS}AUTH_LINK_FAILURE`);
reducer.on(authLinkFailure, state => ({
  ...state,
  isAuthLinkRequesting: false,
  url: null,
  data: null,
}));

const loginUserRequest = createAction(`${NS}LOGIN_USER_REQUEST`);
reducer.on(loginUserRequest, state => ({
  ...state,
  isRequesting: true,
}));

const authLinkRequest = createAction(`${NS}AUTH_LINK_REQUEST`);
reducer.on(authLinkRequest, state => ({
  ...state,
  isAuthLinkRequesting: true,
  url: null,
  data: null,
}));

const reset = createAction(`${NS}RESET`);
reducer.on(reset, () => ({
  ...initialState,
}));

export const getAuthLink = (): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(authLinkRequest());
    GET(
      {
        url: `${EndpointsConfig.AUTH.AUTH_LINK}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        const { url, data } = resp.data as { url: string; data: object };
        // WIP does not use Alicanto.API.GET_AUTH_OAUTH_PROVIDER_URL_RESPONSE for develop
        dispatch(authLinkSuccess({ url, data }));
        resolve();
      })
      .catch(error => {
        dispatch(authLinkFailure());
        reject(error);
      });
  });
};

export const loginUser = (username: string, password: string): Store.IReduxThunkCb => (
  dispatch,
  getState,
  xShopHeaders,
) => {
  return new Promise((resolve, reject) => {
    dispatch(loginUserRequest());
    const data: Alicanto.API.POST_AUTH_LOGIN_BODY = {
      username,
      password,
    };
    POST(
      {
        data,
        url: `${EndpointsConfig.AUTH.LOGIN}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        dispatch(loginUserSuccess());
        if (process.env.BROWSER) {
          // TODO: move to external
          // redirect to home page
          dispatch(push('/'));
        }
        resolve();
      })
      .catch(error => {
        dispatch(loginUserFailure());
        reject(error);
      });
  });
};

export const logoutUser = (cb: () => void): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(loginUserRequest());
    GET(
      {
        url: `${EndpointsConfig.AUTH.LOGOUT}`,
      },
      xShopHeaders,
    )
      .then((resp: AxiosResponse) => {
        dispatch(loginUserSuccess());
        dispatch(reset());
        if (cb) {
          cb();
        }
        if (process.env.BROWSER) {
          // TODO: move to external
          // redirect to home page
          dispatch(push('/'));
        }
        resolve();
      })
      .catch(error => {
        dispatch(loginUserFailure());
        reject(error);
      });
  });
};

export default reducer;
