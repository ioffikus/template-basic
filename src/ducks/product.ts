import { createAction, createReducer } from 'redux-act';
import { AxiosResponse } from 'axios';
import { matchPath } from 'react-router';
import _ from 'lodash';
import { GET } from 'src/core/API';
import EndpointsConfig from 'src/configs/endpoints';
import { setMeta } from 'src/ducks/meta';

export const REDUCER = 'PRODUCTS';
const NS = `${REDUCER}__`;

export interface IProductState extends Alicanto.API.GET_CATALOG_UID_RESPONSE, Store.IReduxCommonState {}

export const initialState: IProductState = {
  uid: null,
  id: null,
  is_unlimited: null,
  schema_fields: {},
  categories: null,
  modifications: null,
  pictures: null,
  isRequesting: false,
  isRequestError: false,
};

const reducer = createReducer<IProductState>({}, initialState);

const callSuccess = createAction<Alicanto.API.GET_CATALOG_UID_RESPONSE>(`${NS}CALL_SUCCESS`);
reducer.on(callSuccess, (state, { id, uid, schema_fields, categories, modifications, pictures, is_unlimited }) => ({
  ...state,
  uid,
  id: _.isNil(id) ? null : id,
  is_unlimited: _.isNil(is_unlimited) ? null : is_unlimited,
  schema_fields: { ...schema_fields },
  categories: _.isNil(categories) ? null : [...categories],
  modifications: _.isNil(modifications) ? null : [...modifications],
  pictures: _.isNil(pictures) ? null : [...pictures],
  isRequesting: false,
  isRequestError: false,
}));

const callFailure = createAction(`${NS}CALL_FAILURE`);
reducer.on(callFailure, state => ({
  ...state,
  uid: null,
  id: null,
  is_unlimited: null,
  schema_fields: {},
  categories: null,
  modifications: null,
  pictures: null,
  isRequestError: true,
}));

const callRequest = createAction(`${NS}CALL_REQUEST`);
reducer.on(callRequest, state => ({
  ...state,
  isRequesting: true,
}));

export const getProduct = (abortToken: string): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(callRequest());
    const router = getState().router;
    const match = matchPath(router.location.pathname, { path: '/product/:uid' });
    const uid = _.get(match, 'params.uid', null);
    GET(
      {
        url: `${EndpointsConfig.PRODUCT}${uid}`,
      },
      xShopHeaders,
      abortToken,
    )
      .then((resp: AxiosResponse) => {
        const data = resp.data as Alicanto.API.GET_CATALOG_UID_RESPONSE;
        dispatch(callSuccess(data));
        resolve();
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  }).then(() => {
    const schemaFields = getState().product.schema_fields;
    const name = _.get(schemaFields, 'name.value', '');
    const description = _.get(schemaFields, 'description.value', '');
    dispatch(setMeta({ description, title: name, keywords: `${name} ${description}` }));
  });
};

export default reducer;
