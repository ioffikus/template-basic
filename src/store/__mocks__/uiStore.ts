import { IUIState } from 'src/ducks/ui';
import fs from 'fs';
import yaml from 'js-yaml';

const globalConfig = yaml.safeLoad(fs.readFileSync('src/configs/global.yml', 'utf8'));
export const ui = globalConfig.ui;

const state: IUIState = {
  ...ui,
  isRequestError: false,
  isRequesting: false,
};

export default state;
