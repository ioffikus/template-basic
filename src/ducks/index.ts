// instead this, we have router in store via connected-react-router
// import { routerReducer as routing, RouterState } from 'react-router-redux';
import { combineReducers } from 'redux';
import auth from './auth';
import cart from './cart';
import cartPaidItems from './cartPaidItems';
import catalogQuery from './catalogQuery';
import catalog from './catalog';
import categories from './categories';
import categoriesMenu from './categoriesMenu';
import filter from './filter';
import locale from './locale';
import meta from './meta';
import page from './page';
import product from './product';
import progress from './progress';
import popularItems from './popularItems';
import recommendItems from './recommendItems';
import routes from './routes';
import ui from './ui';
import user from './user';
import userDelivery from './userDelivery';
import userOrders from './userOrders';

export default combineReducers<Store.IState>({
  ui,
  auth,
  user,
  categories,
  categoriesMenu,
  catalogQuery,
  catalog,
  product,
  progress,
  routes,
  meta,
  filter,
  cart,
  cartPaidItems,
  userDelivery,
  userOrders,
  popularItems,
  recommendItems,
  page,
  locale,
});
