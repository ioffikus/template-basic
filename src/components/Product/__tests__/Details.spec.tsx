import React from 'react';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import { Tabs } from 'antd';
import uiStore from 'src/store/__mocks__/uiStore';
import productStore from 'src/store/__mocks__/productStore';
import { initialState as productInitialState } from 'src/ducks/product';
import Details from '../Details';

const mockStore = configureMockStore([thunk]);

describe(null, () => {
  it('render() with initialState', () => {
    const store = mockStore({
      ui: uiStore,
      product: productInitialState,
    });
    const wrapper = shallow(<Details />, { context: { store } });
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('render() with product', () => {
    const store = mockStore({
      ui: uiStore,
      product: productStore,
    });
    const wrapper = shallow(<Details />, { context: { store } });
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('change active tab', () => {
    const store = mockStore({
      ui: uiStore,
      product: productStore,
    });
    const wrapper = shallow(<Details />, { context: { store } }).dive();
    expect(wrapper).toMatchSnapshot();
    const tabsInstance = wrapper
      .find(Tabs)
      .dive()
      .instance() as Tabs;
    tabsInstance.props.onChange('modifications');
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
});
