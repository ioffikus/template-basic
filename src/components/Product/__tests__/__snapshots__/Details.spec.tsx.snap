// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change active tab 1`] = `
<div
  className="jsx-392034345 root"
>
  <JSXStyle
    css=".root.jsx-392034345 .ant-tabs-bar{text-transform:uppercase;}.root.jsx-392034345 .ant-tabs-tabpane{padding-left:26px;padding-right:26px;}.root.jsx-392034345 .ant-table{font-size:14px;}.root.jsx-392034345 .ant-table-tbody > tr > td{padding-top:17px;padding-bottom:17px;}.root.jsx-392034345 .ant-table td:first-of-type{color:#b2bdcb;}.root.jsx-392034345 .ant-table tr:hover > td{background:transparent;}li.jsx-392034345{padding-top:17px;padding-bottom:17px;}.root.jsx-392034345 .bundle,.root.jsx-392034345 .modification,.root.jsx-392034345 .category{font-size:14px;}.root.jsx-392034345 .modification .currency{font-size:inherit;color:inherit;}"
    styleId="392034345"
  />
  <Tabs
    activeKey="specifications"
    hideAdd={false}
    onChange={[Function]}
    prefixCls="ant-tabs"
  >
    <TabPane
      key="specifications"
      placeholder={null}
      tab="Specifications"
    >
      <Table
        bordered={false}
        className=""
        columns={
          Array [
            Object {
              "dataIndex": "label",
            },
            Object {
              "dataIndex": "value",
            },
          ]
        }
        dataSource={
          Array [
            Object {
              "key": "shop_url",
              "label": "Current Shop URL",
              "value": "template-basic.agg.loc",
            },
            Object {
              "key": "author",
              "label": "Автор",
              "value": "Robert A. Heinlein",
            },
            Object {
              "key": "isbn",
              "label": "ISBN",
              "value": "978-0441783588",
            },
            Object {
              "key": "ebook_type",
              "label": "Тип eBook",
              "value": "PDF",
            },
          ]
        }
        indentSize={20}
        loading={false}
        locale={Object {}}
        pagination={false}
        prefixCls="ant-table"
        rowKey="key"
        rowSelection={null}
        showHeader={false}
        size="large"
        useFixedHeader={false}
      />
    </TabPane>
    <TabPane
      key="modifications"
      placeholder={null}
      tab="Modifications"
    >
      <ul
        className="jsx-392034345"
      >
        <li
          className="jsx-392034345"
          key="58738032e7e0461ca6bb"
        >
          <Link
            className="modification"
            replace={false}
            title="ePub version of Starship Troopers ru"
            to="/product/58738032e7e0461ca6bb"
          >
            Starship Troopers. ePub ru
             - 
            <Connect(Price)
              currency="€"
              price={12}
            />
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="c1b2170baf484f548362"
        >
          <Link
            className="modification"
            replace={false}
            title="Kindle version of Starship Troopers ru"
            to="/product/c1b2170baf484f548362"
          >
            Starship Troopers. Kindle ru
             - 
            <Connect(Price)
              currency="€"
              price={9}
            />
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="5a8343812fd1403aa95f"
        >
          <Link
            className="modification"
            replace={false}
            title="ePub version of Starship Troopers"
            to="/product/5a8343812fd1403aa95f"
          >
            Starship Troopers. ePub. Free!
             - 
            <Connect(Price)
              currency="€"
              price={0}
            />
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="717b27127f7449dd9631"
        >
          <Link
            className="modification"
            replace={false}
            title="PDF version of Starship Troopers"
            to="/product/717b27127f7449dd9631"
          >
            Starship Troopers. PDF. Free!
             - 
            <Connect(Price)
              currency="€"
              price={0}
            />
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="73f09f43bfde4b669c96"
        >
          <Link
            className="modification"
            replace={false}
            title="Kindle version of Starship Troopers"
            to="/product/73f09f43bfde4b669c96"
          >
            Starship Troopers. Kindle. Free!
             - 
            <Connect(Price)
              currency="€"
              price={0}
            />
          </Link>
        </li>
      </ul>
    </TabPane>
    <TabPane
      key="categories"
      placeholder={null}
      tab="Categories"
    >
      <ul
        className="jsx-392034345"
      >
        <li
          className="jsx-392034345"
          key="7911d4c525c74c889309"
        >
          <Link
            className="category"
            replace={false}
            to="/catalog/7911d4c525c74c889309"
          >
            books
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="da26e5a32a12405a9a54"
        >
          <Link
            className="category"
            replace={false}
            to="/catalog/da26e5a32a12405a9a54"
          >
            eBooks
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="c55784279bc64422aeaf"
        >
          <Link
            className="category"
            replace={false}
            to="/catalog/c55784279bc64422aeaf"
          >
            PDF
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="13be1baa4e084ebfa207"
        >
          <Link
            className="category"
            replace={false}
            to="/catalog/13be1baa4e084ebfa207"
          >
            Speculative fiction
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="00643bfe21a148b485ca"
        >
          <Link
            className="category"
            replace={false}
            to="/catalog/00643bfe21a148b485ca"
          >
            Science fiction
          </Link>
        </li>
      </ul>
    </TabPane>
  </Tabs>
</div>
`;

exports[`change active tab 2`] = `
<div
  className="jsx-392034345 root"
>
  <JSXStyle
    css=".root.jsx-392034345 .ant-tabs-bar{text-transform:uppercase;}.root.jsx-392034345 .ant-tabs-tabpane{padding-left:26px;padding-right:26px;}.root.jsx-392034345 .ant-table{font-size:14px;}.root.jsx-392034345 .ant-table-tbody > tr > td{padding-top:17px;padding-bottom:17px;}.root.jsx-392034345 .ant-table td:first-of-type{color:#b2bdcb;}.root.jsx-392034345 .ant-table tr:hover > td{background:transparent;}li.jsx-392034345{padding-top:17px;padding-bottom:17px;}.root.jsx-392034345 .bundle,.root.jsx-392034345 .modification,.root.jsx-392034345 .category{font-size:14px;}.root.jsx-392034345 .modification .currency{font-size:inherit;color:inherit;}"
    styleId="392034345"
  />
  <Tabs
    activeKey="modifications"
    hideAdd={false}
    onChange={[Function]}
    prefixCls="ant-tabs"
  >
    <TabPane
      key="specifications"
      placeholder={null}
      tab="Specifications"
    >
      <Table
        bordered={false}
        className=""
        columns={
          Array [
            Object {
              "dataIndex": "label",
            },
            Object {
              "dataIndex": "value",
            },
          ]
        }
        dataSource={
          Array [
            Object {
              "key": "shop_url",
              "label": "Current Shop URL",
              "value": "template-basic.agg.loc",
            },
            Object {
              "key": "author",
              "label": "Автор",
              "value": "Robert A. Heinlein",
            },
            Object {
              "key": "isbn",
              "label": "ISBN",
              "value": "978-0441783588",
            },
            Object {
              "key": "ebook_type",
              "label": "Тип eBook",
              "value": "PDF",
            },
          ]
        }
        indentSize={20}
        loading={false}
        locale={Object {}}
        pagination={false}
        prefixCls="ant-table"
        rowKey="key"
        rowSelection={null}
        showHeader={false}
        size="large"
        useFixedHeader={false}
      />
    </TabPane>
    <TabPane
      key="modifications"
      placeholder={null}
      tab="Modifications"
    >
      <ul
        className="jsx-392034345"
      >
        <li
          className="jsx-392034345"
          key="58738032e7e0461ca6bb"
        >
          <Link
            className="modification"
            replace={false}
            title="ePub version of Starship Troopers ru"
            to="/product/58738032e7e0461ca6bb"
          >
            Starship Troopers. ePub ru
             - 
            <Connect(Price)
              currency="€"
              price={12}
            />
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="c1b2170baf484f548362"
        >
          <Link
            className="modification"
            replace={false}
            title="Kindle version of Starship Troopers ru"
            to="/product/c1b2170baf484f548362"
          >
            Starship Troopers. Kindle ru
             - 
            <Connect(Price)
              currency="€"
              price={9}
            />
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="5a8343812fd1403aa95f"
        >
          <Link
            className="modification"
            replace={false}
            title="ePub version of Starship Troopers"
            to="/product/5a8343812fd1403aa95f"
          >
            Starship Troopers. ePub. Free!
             - 
            <Connect(Price)
              currency="€"
              price={0}
            />
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="717b27127f7449dd9631"
        >
          <Link
            className="modification"
            replace={false}
            title="PDF version of Starship Troopers"
            to="/product/717b27127f7449dd9631"
          >
            Starship Troopers. PDF. Free!
             - 
            <Connect(Price)
              currency="€"
              price={0}
            />
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="73f09f43bfde4b669c96"
        >
          <Link
            className="modification"
            replace={false}
            title="Kindle version of Starship Troopers"
            to="/product/73f09f43bfde4b669c96"
          >
            Starship Troopers. Kindle. Free!
             - 
            <Connect(Price)
              currency="€"
              price={0}
            />
          </Link>
        </li>
      </ul>
    </TabPane>
    <TabPane
      key="categories"
      placeholder={null}
      tab="Categories"
    >
      <ul
        className="jsx-392034345"
      >
        <li
          className="jsx-392034345"
          key="7911d4c525c74c889309"
        >
          <Link
            className="category"
            replace={false}
            to="/catalog/7911d4c525c74c889309"
          >
            books
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="da26e5a32a12405a9a54"
        >
          <Link
            className="category"
            replace={false}
            to="/catalog/da26e5a32a12405a9a54"
          >
            eBooks
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="c55784279bc64422aeaf"
        >
          <Link
            className="category"
            replace={false}
            to="/catalog/c55784279bc64422aeaf"
          >
            PDF
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="13be1baa4e084ebfa207"
        >
          <Link
            className="category"
            replace={false}
            to="/catalog/13be1baa4e084ebfa207"
          >
            Speculative fiction
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="00643bfe21a148b485ca"
        >
          <Link
            className="category"
            replace={false}
            to="/catalog/00643bfe21a148b485ca"
          >
            Science fiction
          </Link>
        </li>
      </ul>
    </TabPane>
  </Tabs>
</div>
`;

exports[`render() with initialState 1`] = `
<div
  className="jsx-392034345 root"
>
  <JSXStyle
    css=".root.jsx-392034345 .ant-tabs-bar{text-transform:uppercase;}.root.jsx-392034345 .ant-tabs-tabpane{padding-left:26px;padding-right:26px;}.root.jsx-392034345 .ant-table{font-size:14px;}.root.jsx-392034345 .ant-table-tbody > tr > td{padding-top:17px;padding-bottom:17px;}.root.jsx-392034345 .ant-table td:first-of-type{color:#b2bdcb;}.root.jsx-392034345 .ant-table tr:hover > td{background:transparent;}li.jsx-392034345{padding-top:17px;padding-bottom:17px;}.root.jsx-392034345 .bundle,.root.jsx-392034345 .modification,.root.jsx-392034345 .category{font-size:14px;}.root.jsx-392034345 .modification .currency{font-size:inherit;color:inherit;}"
    styleId="392034345"
  />
  <Tabs
    activeKey="specifications"
    hideAdd={false}
    onChange={[Function]}
    prefixCls="ant-tabs"
  >
    <TabPane
      key="specifications"
      placeholder={null}
      tab="Specifications"
    >
      <Table
        bordered={false}
        className=""
        columns={
          Array [
            Object {
              "dataIndex": "label",
            },
            Object {
              "dataIndex": "value",
            },
          ]
        }
        dataSource={Array []}
        indentSize={20}
        loading={false}
        locale={Object {}}
        pagination={false}
        prefixCls="ant-table"
        rowKey="key"
        rowSelection={null}
        showHeader={false}
        size="large"
        useFixedHeader={false}
      />
    </TabPane>
  </Tabs>
</div>
`;

exports[`render() with product 1`] = `
<div
  className="jsx-392034345 root"
>
  <JSXStyle
    css=".root.jsx-392034345 .ant-tabs-bar{text-transform:uppercase;}.root.jsx-392034345 .ant-tabs-tabpane{padding-left:26px;padding-right:26px;}.root.jsx-392034345 .ant-table{font-size:14px;}.root.jsx-392034345 .ant-table-tbody > tr > td{padding-top:17px;padding-bottom:17px;}.root.jsx-392034345 .ant-table td:first-of-type{color:#b2bdcb;}.root.jsx-392034345 .ant-table tr:hover > td{background:transparent;}li.jsx-392034345{padding-top:17px;padding-bottom:17px;}.root.jsx-392034345 .bundle,.root.jsx-392034345 .modification,.root.jsx-392034345 .category{font-size:14px;}.root.jsx-392034345 .modification .currency{font-size:inherit;color:inherit;}"
    styleId="392034345"
  />
  <Tabs
    activeKey="specifications"
    hideAdd={false}
    onChange={[Function]}
    prefixCls="ant-tabs"
  >
    <TabPane
      key="specifications"
      placeholder={null}
      tab="Specifications"
    >
      <Table
        bordered={false}
        className=""
        columns={
          Array [
            Object {
              "dataIndex": "label",
            },
            Object {
              "dataIndex": "value",
            },
          ]
        }
        dataSource={
          Array [
            Object {
              "key": "shop_url",
              "label": "Current Shop URL",
              "value": "template-basic.agg.loc",
            },
            Object {
              "key": "author",
              "label": "Автор",
              "value": "Robert A. Heinlein",
            },
            Object {
              "key": "isbn",
              "label": "ISBN",
              "value": "978-0441783588",
            },
            Object {
              "key": "ebook_type",
              "label": "Тип eBook",
              "value": "PDF",
            },
          ]
        }
        indentSize={20}
        loading={false}
        locale={Object {}}
        pagination={false}
        prefixCls="ant-table"
        rowKey="key"
        rowSelection={null}
        showHeader={false}
        size="large"
        useFixedHeader={false}
      />
    </TabPane>
    <TabPane
      key="modifications"
      placeholder={null}
      tab="Modifications"
    >
      <ul
        className="jsx-392034345"
      >
        <li
          className="jsx-392034345"
          key="58738032e7e0461ca6bb"
        >
          <Link
            className="modification"
            replace={false}
            title="ePub version of Starship Troopers ru"
            to="/product/58738032e7e0461ca6bb"
          >
            Starship Troopers. ePub ru
             - 
            <Connect(Price)
              currency="€"
              price={12}
            />
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="c1b2170baf484f548362"
        >
          <Link
            className="modification"
            replace={false}
            title="Kindle version of Starship Troopers ru"
            to="/product/c1b2170baf484f548362"
          >
            Starship Troopers. Kindle ru
             - 
            <Connect(Price)
              currency="€"
              price={9}
            />
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="5a8343812fd1403aa95f"
        >
          <Link
            className="modification"
            replace={false}
            title="ePub version of Starship Troopers"
            to="/product/5a8343812fd1403aa95f"
          >
            Starship Troopers. ePub. Free!
             - 
            <Connect(Price)
              currency="€"
              price={0}
            />
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="717b27127f7449dd9631"
        >
          <Link
            className="modification"
            replace={false}
            title="PDF version of Starship Troopers"
            to="/product/717b27127f7449dd9631"
          >
            Starship Troopers. PDF. Free!
             - 
            <Connect(Price)
              currency="€"
              price={0}
            />
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="73f09f43bfde4b669c96"
        >
          <Link
            className="modification"
            replace={false}
            title="Kindle version of Starship Troopers"
            to="/product/73f09f43bfde4b669c96"
          >
            Starship Troopers. Kindle. Free!
             - 
            <Connect(Price)
              currency="€"
              price={0}
            />
          </Link>
        </li>
      </ul>
    </TabPane>
    <TabPane
      key="categories"
      placeholder={null}
      tab="Categories"
    >
      <ul
        className="jsx-392034345"
      >
        <li
          className="jsx-392034345"
          key="7911d4c525c74c889309"
        >
          <Link
            className="category"
            replace={false}
            to="/catalog/7911d4c525c74c889309"
          >
            books
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="da26e5a32a12405a9a54"
        >
          <Link
            className="category"
            replace={false}
            to="/catalog/da26e5a32a12405a9a54"
          >
            eBooks
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="c55784279bc64422aeaf"
        >
          <Link
            className="category"
            replace={false}
            to="/catalog/c55784279bc64422aeaf"
          >
            PDF
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="13be1baa4e084ebfa207"
        >
          <Link
            className="category"
            replace={false}
            to="/catalog/13be1baa4e084ebfa207"
          >
            Speculative fiction
          </Link>
        </li>
        <li
          className="jsx-392034345"
          key="00643bfe21a148b485ca"
        >
          <Link
            className="category"
            replace={false}
            to="/catalog/00643bfe21a148b485ca"
          >
            Science fiction
          </Link>
        </li>
      </ul>
    </TabPane>
  </Tabs>
</div>
`;
